File: runme/mcgen/local/datan/datsrc/mingld.c
#include "datsrc.h"

void LIBFUNC mingld_(double P_T *a, double P_T *b, double P_T *c,
             double const P_T *epsiln,
             double (CBFUNC P_T *userfn)(double const P_T *,
             double const P_T *, double const P_T *, integer const P_T *),
             double P_T *xmin, double P_T *ymin, integer P_T *nstep,
             double const P_T *x0, double const P_T *xdir,
             integer const P_T *n)
{
    /* System generated locals */
    double d__1, d__2;

    /* Local variables */
    integer istep;
    double s, x, y, eps;

    eps = *epsiln;
    if (eps <= 0.f) {
        eps = 1e-8;
    }
    if (*nstep < 1) {
        *nstep = 1000;
    }
    s = (*userfn)(b, x0, xdir, n);
    istep = 0;
L10:
    ++istep;
    if ((d__1 = *b - *a, abs(d__1)) > (d__2 = *c - *b, abs(d__2))) {
        x = *a + (*b - *a) * .61803;
        y = (*userfn)(&x, x0, xdir, n);
        if (y < s) {
            *c = *b;
            *b = x;
            s = y;
        } else {
            *a = x;
        }
    } else {
        x = *b + (*c - *b) * .61803;
        y = (*userfn)(&x, x0, xdir, n);
        if (y < s) {
            *a = *b;
            *b = x;
            s = y;
        } else {
            *c = x;
        }
    }
    if ((d__1 = *c - *a, abs(d__1)) > eps * abs(*a) + 1e-15 &&
          istep <= *nstep) {
        goto L10;
    }
    if (istep > *nstep) {
        *nstep = -1;
    } else {
        *nstep = istep;
    }
    *xmin = *b;
    *ymin = *c;
    return;
} /* mingld_ */

File: runme/mcgen/local/datan/datsrc/mingld.f
      SUBROUTINE MINGLD(A,B,C,EPSILN,USERFN,XMIN,YMIN,NSTEP,X0,
     +XDIR,N)
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(Q=0.61803D0,MAXSTP=1000,EPSDEF=1.D-8,TT=1.D-15)
      EXTERNAL USERFN
      EPS=EPSILN
      IF(EPS.LE.0.)EPS=EPSDEF
      IF(NSTEP.LT.1)NSTEP=MAXSTP
      S=USERFN(B,X0,XDIR,N)
      ISTEP=0
   10 ISTEP=ISTEP+1
      IF(ABS(B-A).GT.ABS(C-B)) THEN
        X=A+Q*(B-A)
        Y=USERFN(X,X0,XDIR,N)
        IF(Y.LT.S) THEN
          C=B
          B=X
          S=Y
        ELSE
          A=X
        END IF
      ELSE
        X=B+Q*(C-B)
        Y=USERFN(X,X0,XDIR,N)
        IF(Y.LT.S) THEN
          A=B
          B=X
          S=Y
        ELSE
          C=X
        END IF
      END IF
      IF (ABS(C-A).GT.EPS*ABS(A)+TT .AND. ISTEP.LE.NSTEP) GO TO 10
      IF(ISTEP.GT.NSTEP) THEN
        NSTEP=-1
      ELSE
        NSTEP=ISTEP
      END IF
      XMIN=B
      YMIN=C
      END
